# Capstone project Week 3 assignment

##  Part 1

### Import required libraries

In [1]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge lxml --yes 
!pip install et_xmlfile

#!conda install -c conda-forge folium=0.11.0 --yes #installed 19jun20
import folium # plotting library


Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.



#### Read the table from the Wikipedia page directly by pandas

In [2]:
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wikitable = pd.read_html(url)

wikitable

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                           Neighborhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

#### It can be seen from the html output that the postal code table is in the [0] position.
#### A pandas data frame is created from position [0] of wikitable. 

In [3]:
df_postalCode=pd.DataFrame(wikitable[0])

df_postalCode

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

#### Remove the boroughs that are "Not Assigned" and check if there will be any "NaN" neighborhood remaining.

In [5]:
notAssignedB = df_postalCode['Borough'] == 'Not assigned'
notAssignedN = df_postalCode['Neighborhood'] == 'Not assigned'
notAssignedIndexB= df_postalCode[notAssignedB].index
notAssignedIndexN= df_postalCode[notAssignedN].index
if all(notAssignedIndexB == notAssignedIndexN):
    df_postalCode.drop(notAssignedIndexB, inplace=True)
    print("None unassigned neighborhoods remain")

None unassigned neighborhoods remain


#### There are no neighborhoods to be renamed after removing the "Not Assigned" boroughs.

In [6]:
print(df_postalCode.shape)
df_postalCode

(103, 3)


Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

#### The resulting data frame has 103 postal codes (rows).
#### Checking if any Postal Code is repeated; if yes, the corresponding neighborhoods will need to be merged.

In [7]:
df_postalCode.nunique()

Postal Code     103
Borough          10
Neighborhood     99
dtype: int64

#### There are 103 different postal codes, which means there are no repeats.
#### There are ten different boroughs in the 103 postal codes. It was known in advance that boroughs are repeated.
#### There are four repeated neighborhoods. Which ones?

In [8]:
dups=df_postalCode['Neighborhood'].duplicated()
df_postalCode[dups].sort_values(by=df_postalCode.columns.to_list())

Postal Code     Borough Neighborhood
20         M3C  North York    Don Mills
74         M3L  North York    Downsview
83         M3M  North York    Downsview
92         M3N  North York    Downsview

In [9]:
dups2 = df_postalCode['Neighborhood'] == 'Don Mills'
dups3 = df_postalCode['Neighborhood'] == 'Downsview'
print(df_postalCode[dups2].sort_values(by=df_postalCode.columns.to_list()),'\n')
print(df_postalCode[dups3].sort_values(by=df_postalCode.columns.to_list()))

   Postal Code     Borough Neighborhood
11         M3B  North York    Don Mills
20         M3C  North York    Don Mills 

   Postal Code     Borough Neighborhood
65         M3K  North York    Downsview
74         M3L  North York    Downsview
83         M3M  North York    Downsview
92         M3N  North York    Downsview


#### According to the repeats lists, Don Mills appears twice in the data frame and Downsview appears four times.
#### However, there is indeed a different postal code for every repeated neighborhood.
#### This means that there is nothing more to clean up in this data set and there are 103 different postal codes in Toronto with an assigned borough and neighborhood.

In [10]:
print(df_postalCode.shape)

df_postalCode

(103, 3)


Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]